## Exercises: Filter and query data

In [2]:
import pandas as pd
autos = pd.read_json("../Data/autos.json")
autos

aspiration   body-style  bore  city-mpg  compression-ratio  curb-weight  \
0          std  convertible  3.47        21                9.0         2548   
1          std  convertible  3.47        21                9.0         2548   
2          std    hatchback  2.68        19                9.0         2823   
3          std        sedan  3.19        24               10.0         2337   
4          std        sedan  3.19        18                8.0         2824   
..         ...          ...   ...       ...                ...          ...   
200        std        sedan  3.78        23                9.5         2952   
201      turbo        sedan  3.78        19                8.7         3049   
202        std        sedan  3.58        18                8.8         3012   
203      turbo        sedan  3.01        26               23.0         3217   
204      turbo        sedan  3.78        19                9.5         3062   

    drive-wheels engine-location  engine-size engine-type  ...         make  \
0            rwd           front          130        dohc  ...  alfa-romero   
1            rwd           front          130        dohc  ...  alfa-romero   
2            rwd           front          152        ohcv  ...  alfa-romero   
3            fwd           front          109         ohc  ...         audi   
4            4wd           front          136         ohc  ...         audi   
..           ...             ...          ...         ...  ...          ...   
200          rwd           front          141         ohc  ...        volvo   
201          rwd           front          141         ohc  ...        volvo   
202          rwd           front          173        ohcv  ...        volvo   
203          rwd           front          145         ohc  ...        volvo   
204          rwd           front          141         ohc  ...        volvo   

    normalized-losses  num-of-cylinders  num-of-doors  peak-rpm    price  \
0                 NaN              four           two    5000.0  13495.0   
1                 NaN              four           two    5000.0  16500.0   
2                 NaN               six           two    5000.0  16500.0   
3               164.0              four          four    5500.0  13950.0   
4               164.0              five          four    5500.0  17450.0   
..                ...               ...           ...       ...      ...   
200              95.0              four          four    5400.0  16845.0   
201              95.0              four          four    5300.0  19045.0   
202              95.0               six          four    5500.0  21485.0   
203              95.0               six          four    4800.0  22470.0   
204              95.0              four          four    5400.0  22625.0   

    stroke  symboling wheel-base width  
0     2.68          3       88.6  64.1  
1     2.68          3       88.6  64.1  
2     3.47          1       94.5  65.5  
3     3.40          2       99.8  66.2  
4     3.40          2       99.4  66.4  
..     ...        ...        ...   ...  
200   3.15         -1      109.1  68.9  
201   3.15         -1      109.1  68.8  
202   2.87         -1      109.1  68.9  
203   3.40         -1      109.1  68.9  
204   3.15         -1      109.1  68.9  

[205 rows x 26 columns]

**Use the "autos" dataset to answer the following questions:**

**Q1:** How many cars from Jaguar are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=3
&nbsp;&nbsp;&nbsp;price=36000.0
</details>

In [3]:
jaguars = autos.query("make == 'jaguar'")

if not jaguars.empty:
    print(f"Number of Jaguar in dataset: {len(jaguars)}")
    most_expensive_jaguar = jaguars.sort_values(by="price", ascending=False).iloc[0]
    print(f"Most expensive Jaguar: {most_expensive_jaguar['price']}")
else:
    print("No Jaguar cars found in the dataset.")


Number of Jaguar in dataset: 3
Most expensive Jaguar: 36000.0


**Q2:** How many cars from Toyota are in the dataset, and what is the price of the most expensive one?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;count=32
&nbsp;&nbsp;&nbsp;price=17669.0
</details>

In [4]:
toyotas = autos.query("make == 'toyota'")

if not toyotas.empty:
    print(f"Number of toyota in dataset: {len(toyotas)}")
    most_expensive_toyota = toyotas.sort_values(by="price", ascending=False).iloc[0]
    print(f"Most expensive toyota: {most_expensive_toyota['price']}")
else:
    print("No toyota cars found in the dataset.")

Number of toyota in dataset: 32
Most expensive toyota: 17669.0


**Q3:** What is the length, width and height of the most expensive car in the entire dataset?

<details>
<summary>Answer</summary>
<br>&nbsp;&nbsp;&nbsp;length=199.2
&nbsp;&nbsp;&nbsp;width=72.0
&nbsp;&nbsp;&nbsp;height=55.4
</details>

In [5]:
most_expensive_car = autos.sort_values(by="price", ascending=False).iloc[0]

print("Length, width and height of most expensive car in dataset")
print(f"length: {most_expensive_car['length']}, width: {most_expensive_car['width']}, height: {most_expensive_car['height']}")

Length, width and height of most expensive car in dataset
length: 199.2, width: 72.0, height: 55.4


**Q4:** What is the lowest price per horsepower in the dataset, and what brand ("make") is that car?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;price per horsepower=72.84
&nbsp;&nbsp;&nbsp;brand=Toyota
</details>

In [6]:
autos['hp_per_price'] = autos['horsepower'] / autos['price']

max_hp_per_price = autos.sort_values(by="hp_per_price", ascending=False).iloc[0]

print(f"The car with max horsepower per investment is a {max_hp_per_price['make']}, with {max_hp_per_price['horsepower']} HP to a cost of {max_hp_per_price['price']} USD.")
print(f"This gives price per horsepower: {max_hp_per_price['price'] / max_hp_per_price['horsepower']:.2f} USD")

The car with max horsepower per investment is a toyota, with 116.0 HP to a cost of 8449.0 USD.
This gives price per horsepower: 72.84 USD


**Q5:** How many of the cars in the dataset has as many cylinders as they have doors?

<details>
<summary>Answer</summary>
<br>
&nbsp;&nbsp;&nbsp;cars=95
</details>

In [7]:
# Back-ticks if strange charachters in column name
autos_as_many_doors_as_cylinders = autos.query("`num-of-cylinders` == `num-of-doors`")

print(f"Number of cars with as many cylinders as doors: {len(autos_as_many_doors_as_cylinders)}")

Number of cars with as many cylinders as doors: 95


**T1:** Print a string to inform the user of the price difference between the cheapest and the most expensive car in the dataset.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;The price difference between the cheapest and the most expensive car is 40282.0
</details>

In [8]:
most_expensive_car = autos.sort_values(by="price", ascending=False).iloc[0]
least_expensive_car = autos.sort_values(by="price", ascending=True).iloc[0]

print(f"The price difference between the most and least expensive cars is: {most_expensive_car['price'] - least_expensive_car['price']} USD.")


The price difference between the most and least expensive cars is: 40282.0 USD.


**T2:** Ask the user to input a brand, then print the price range for that brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volvo<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'volvo' ranges from 12940.0 to 22625.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: toyota<br>
&nbsp;&nbsp;&nbsp;The prices for cars of brand 'toyota' ranges from 5348.0 to 17669.0<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 3:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

In [9]:
print("Get price range of car brand. Enter emptry string to exit.")
print("==========================================================\n")

while True:
    try:
        brand = input("Enter brand: ").lower()
        if brand == "":
            break
        if len(autos.query("make == @brand")) == 0:
            raise ValueError
        else:
            break
    except ValueError:
        print("Brand not in list of cars. Try again, or enter empty string to exit.")

if brand != "":
    brandcars = autos.query("make == @brand")
    most_expensive_car = brandcars.sort_values(by="price", ascending=False).iloc[0]
    least_expensive_car = brandcars.sort_values(by="price", ascending=True).iloc[0]

    print(f"The price range of {brand.title()} cars is: {least_expensive_car['price']} - {most_expensive_car['price']} USD.")

Get price range of car brand. Enter emptry string to exit.

The price range of Volvo cars is: 12940.0 - 22625.0 USD.


**T3:** Ask the user to input a brand, then print the number of cars in the dataset for that brand, and all attributes for a random sample car of that brand.
<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: mazda<br>
&nbsp;&nbsp;&nbsp;There are 17 cars of brand 'mazda' in the dataset.<br><br>
&nbsp;&nbsp;&nbsp;Here is the data for a random 'mazda' car:<br>
&nbsp;&nbsp;&nbsp;aspiration = std<br>
&nbsp;&nbsp;&nbsp;body-style = sedan<br>
&nbsp;&nbsp;&nbsp;bore = 3.03<br>
&nbsp;&nbsp;&nbsp;city-mpg = 31<br>
&nbsp;&nbsp;&nbsp;compression-ratio = 9.0<br>
&nbsp;&nbsp;&nbsp;curb-weight = 1945<br>
&nbsp;&nbsp;&nbsp;drive-wheels = fwd<br>
&nbsp;&nbsp;&nbsp;engine-location = front<br>
&nbsp;&nbsp;&nbsp;engine-size = 91<br>
&nbsp;&nbsp;&nbsp;engine-type = ohc<br>
&nbsp;&nbsp;&nbsp;fuel-system = 2bbl<br>
&nbsp;&nbsp;&nbsp;fuel-type = gas<br>
&nbsp;&nbsp;&nbsp;height = 54.1<br>
&nbsp;&nbsp;&nbsp;highway-mpg = 38<br>
&nbsp;&nbsp;&nbsp;horsepower = 68.0<br>
&nbsp;&nbsp;&nbsp;length = 166.8<br>
&nbsp;&nbsp;&nbsp;make = mazda<br>
&nbsp;&nbsp;&nbsp;normalized-losses = 113.0<br>
&nbsp;&nbsp;&nbsp;num-of-cylinders = four<br>
&nbsp;&nbsp;&nbsp;num-of-doors = four<br>
&nbsp;&nbsp;&nbsp;peak-rpm = 5000.0<br>
&nbsp;&nbsp;&nbsp;price = 6695.0<br>
&nbsp;&nbsp;&nbsp;stroke = 3.15<br>
&nbsp;&nbsp;&nbsp;symboling = 1<br>
&nbsp;&nbsp;&nbsp;wheel-base = 93.1<br>
&nbsp;&nbsp;&nbsp;width = 64.2<br>
</details>

In [10]:
print("Get number of cars of brand, and all attributes of random car of that brand.\n Enter emptry string to exit.")
print("============================================================================\n")

while True:
    try:
        brand = input("Enter brand: ").lower()
        if brand == "":
            break
        if len(autos.query("make == @brand")) == 0:
            raise ValueError
        else:
            break
    except ValueError:
        print("Brand not in list of cars. Try again, or enter empty string to exit.")

if brand != "":
    brandcars = autos.query("make == @brand")
    print(f"Number of {brand.title()} cars: {len(brandcars)}")
    print(f"All attributes of random {brand.title()} car:\n")
    print(f"{brandcars.sample(1)}")


Get number of cars of brand, and all attributes of random car of that brand.
 Enter emptry string to exit.

Number of Volvo cars: 11
All attributes of random Volvo car:

    aspiration body-style  bore  city-mpg  compression-ratio  curb-weight  \
196        std      sedan  3.78        24                9.5         2935   

    drive-wheels engine-location  engine-size engine-type  ...  \
196          rwd           front          141         ohc  ...   

    normalized-losses num-of-cylinders  num-of-doors  peak-rpm    price  \
196             103.0             four          four    5400.0  15985.0   

     stroke symboling  wheel-base width hp_per_price  
196    3.15        -2       104.3  67.2     0.007132  

[1 rows x 27 columns]


**T4:** Ask the user to input a brand, then export all cars of that brand to a .csv file with the same name as the brand.

<details>
<summary>Solution</summary>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 1:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: volkswagen<br>
&nbsp;&nbsp;&nbsp;Exported 12 cars to 'volkswagen.csv'<br>
<br>
&nbsp;&nbsp;&nbsp;<b>Example 2:</b><br>
&nbsp;&nbsp;&nbsp;Input the name of a brand: tesla<br>
&nbsp;&nbsp;&nbsp;The brand 'tesla' does not exists in the dataset.<br>
</details>

In [11]:
print("Export all cars of brand.\nEnter emptry string to exit.")
print("============================================================================\n")

while True:
    try:
        brand = input("Enter brand: ").strip().lower()
        if brand == "":
            break
        if len(autos.query("make == @brand")) == 0:
            raise ValueError
        else:
            break
    except ValueError:
        print("Brand not in list of cars. Try again, or enter empty string to exit.")

if brand != "":
    brandcars = autos.query("make == @brand")
    try:
        brandcars.to_csv(f"'E001_export_{brand}.csv'")
        print(f"Saved file: E001_export_{brand}.csv")
    except:
        print("Could not save csv file.")
    

Export all cars of brand.
Enter emptry string to exit.



Saved file: E001_export_volvo.csv
